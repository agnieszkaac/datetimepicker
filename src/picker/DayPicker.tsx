import React from "react";
import cx from "classnames";
import moment from "moment";

import { DayPickerProps, DayProps } from "../types";

const Day: React.FunctionComponent<DayProps> = ({
  value,
  selected,
  isToday,
  onPick,
}) => (
  <button
    className={cx("day", { selected: selected, "is-today": isToday })}
    onClick={onPick}
    value={value}
  >
    {value}
  </button>
);

export const DayPicker: React.FunctionComponent<DayPickerProps> = ({
  value,
  monthLength,
  onPick,
}) => (
  <div className="day-picker">
    {Array.from({ length: monthLength }, (e, i) => (
      <Day
        key={i + 1}
        value={(i + 1).toString()}
        selected={value === (i + 1).toString()}
        isToday={moment().date().toString() === value}
        onPick={onPick}
      />
    ))}
  </div>
);
