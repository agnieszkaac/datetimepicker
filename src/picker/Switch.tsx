import React from "react";
import moment, { Moment } from "moment";

import { View } from "./types";

export interface SwitchProps {
  viewDate: Moment;
  view: View;
  onClick: () => void;
}

export const Switch: React.FunctionComponent<SwitchProps> = ({
  viewDate,
  view,
  onClick,
}) => {
  const decadeStart = Math.floor(moment(viewDate).get("year") / 10) * 10;
  return (
    <div className="switch" onClick={onClick}>
      {view === View.Year
        ? `${decadeStart} - ${decadeStart + 11}`
        : viewDate.format(
            view === View.Day ? "MMMM YYYY" : view === View.Month ? "YYYY" : "",
          )}
    </div>
  );
};
