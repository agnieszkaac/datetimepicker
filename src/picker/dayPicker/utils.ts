import moment, { Moment } from "moment";

export const weekDaysNumber = 7; //Always show 7 week days => columns
export const showDaysNumber = 42; //Always show 6 weeks => rows

//dayOfWeek is a number in [0 - 6] value range
//return value should be in [1 - weekDaysNumber] value range
export const prevDaysOffset = (dayOfWeek: number) =>
  dayOfWeek === 0 ? weekDaysNumber : dayOfWeek;

export const viewData = (viewDate: Moment, date?: Moment) => {
  const offset = prevDaysOffset(moment(viewDate).date(1).weekday()) - 1; //0-6
  const prevMonth = moment(viewDate).subtract(1, "M");
  const nextMonth = moment(viewDate).add(1, "M");

  return { offset, prevMonth, nextMonth };
};
