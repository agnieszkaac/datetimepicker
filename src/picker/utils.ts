import moment, { Moment } from "moment";

export enum View {
  Day,
  Month,
  Year,
}

export const getNextView = (dir: -1 | 1, curr: View) => {
  if ((dir > 0 && curr < View.Year) || (dir < 0 && curr > View.Day)) {
    return curr + dir;
  }
  return curr;
};

export const getViewChangeOffset = (dir: number, view: View) =>
  view === View.Year ? dir * 10 : dir;

export const getViewChangeUnit = (view: View) =>
  view === View.Day ? "month" : "year";

export const getViewDate = (dir: number, view: View, viewDate: Moment) => {
  const unit = getViewChangeUnit(view);
  return moment(viewDate)
    .add(getViewChangeOffset(dir, view), unit)
    .startOf(unit);
};
