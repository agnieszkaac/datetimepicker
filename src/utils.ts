import { useEffect } from "react";
import moment, { Moment } from "moment";

import { ValueDate } from "./types";
import { View } from "./picker/types";

export const noop = () => null;

export const parseValueToMoment = (value?: ValueDate): Moment | undefined => {
  if (value instanceof Date || typeof value === "string") {
    return moment(value);
  }
  return value;
};

export const parseMomentToString = (
  value: Moment | undefined,
  format: string,
): string => {
  if (!value) {
    return "";
  }
  return value.format(format);
};

export const getDecadeStart = (date: Moment) =>
  Math.floor(moment(date).get("year") / 10) * 10;

export const togglePickerOpen = (
  pickerOpen: boolean,
  handleClick: (e: MouseEvent) => void,
) => {
  useEffect(() => {
    document.addEventListener("mousedown", handleClick);
    return () => {
      document.removeEventListener("mousedown", handleClick);
    };
  }, [pickerOpen]);
};

export const getDecadeLabel = (date: Moment) => {
  const decadeStart = getDecadeStart(date);
  return `${decadeStart} - ${decadeStart + 11}`;
};

export const getViewLabel = (view: View, viewDate: Moment) =>
  view === View.Year
    ? getDecadeLabel(viewDate)
    : viewDate.format(view === View.Month ? "YYYY" : "MMMM YYYY");

export const configLocale = (
  date: Moment | undefined,
  locale: string,
  dow?: number,
) => {
  date?.locale(locale);
  if (dow) {
    console.log("test", dow);
    moment.updateLocale(locale, {
      week: {
        dow,
        doy: 7 + dow - 1,
      },
    });
  }
};
