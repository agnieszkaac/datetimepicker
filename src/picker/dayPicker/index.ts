import { Day } from "./Day";
import DayPicker from "./DayPicker";
import { DayProps, DayPickerProps } from "./types";

export { Day, DayPicker, DayProps, DayPickerProps };
