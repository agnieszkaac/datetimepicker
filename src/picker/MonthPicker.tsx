import React from "react";
import moment from "moment";
import cx from "classnames";

import { PickerComponentProps } from "./types";
import { defaultGridSize as length } from "./utils";
import { DateContext, ViewContext } from "../state";

import { PickButton } from "./PickButton";

import "./MonthPicker.scss";

export const MonthPicker: React.FC<PickerComponentProps> = ({ onPick }) => {
  const { viewDate } = React.useContext(ViewContext);
  const { date } = React.useContext(DateContext);

  return (
    <div className="month-picker">
      {Array.from({ length }, (_e, i) => {
        const month = moment(viewDate).month(i);
        return (
          <PickButton
            key={i}
            type="month"
            date={month.startOf("month")}
            format="MMM"
            className={cx({
              selected: date?.month() === i && date?.isSame(viewDate, "y"),
              "is-today": month.isSame(moment(), "M"),
            })}
            onClick={onPick}
          />
        );
      })}
    </div>
  );
};
