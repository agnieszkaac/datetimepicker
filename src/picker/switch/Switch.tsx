import React from "react";
import { Moment } from "moment";

import { RangeSwitch } from "./RangeSwitch";
import { ViewSwitch } from "./ViewSwitch";

import { View } from "../utils";
import { getViewLabel } from "../../utils";

import "./Switch.scss";

export type SwitchProps = {
  view: View;
  viewDate: Moment;
  onViewSwitch: () => void;
  onRangeSwitch: (e: React.MouseEvent<HTMLButtonElement>) => void;
}

export const Switch: React.FunctionComponent<SwitchProps> = ({
  view,
  viewDate,
  onViewSwitch,
  onRangeSwitch,
}) => (
  <div className="switch">
    <RangeSwitch value={-1} onClick={onRangeSwitch} />
    <ViewSwitch label={getViewLabel(view, viewDate)} onClick={onViewSwitch} />
    <RangeSwitch value={1} onClick={onRangeSwitch} />
  </div>
);
