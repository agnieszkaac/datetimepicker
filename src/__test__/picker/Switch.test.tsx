import React from "react";
import MockDate from "mockdate";
import moment, { Moment } from "moment";
import { render } from "@testing-library/react";

import { Switch, SwitchProps } from "../../picker/switch/Switch";
import { View } from "../../picker/utils";

describe("<Switch /> should", () => {
  let props: SwitchProps;
  let viewDate: Moment;

  beforeEach(() => {
    MockDate.set("2020-01-01");

    viewDate = moment();
    props = {
      view: View.Day,
      viewDate,
      onViewSwitch: jest.fn(),
      onRangeSwitch: jest.fn(),
    };
  });

  it("match snapshot", () => {
    const { asFragment } = render(<Switch {...props} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
