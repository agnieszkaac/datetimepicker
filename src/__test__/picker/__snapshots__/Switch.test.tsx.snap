// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Switch /> should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="switch"
  >
    <button
      class="range-switch"
      data-testid="range-switch"
      value="-1"
    >
      <span>
        ❮
      </span>
    </button>
    <button
      class="view-switch"
      data-testid="view-switch"
    >
      January 2020
    </button>
    <button
      class="range-switch"
      data-testid="range-switch"
      value="1"
    >
      <span>
        ❯
      </span>
    </button>
  </div>
</DocumentFragment>
`;
